test/pipelining/PipeliningTest.js

```diff
@@ -0,0 +1,237 @@
+        expect(pipelining.run()).to.be.be.rejectedWith('Error1');
+    });
+
+    it('should reject if any of the requests fail with storage of the results', function () {
+        var pipelining = new Pipelining(1, createRejectingLoadGenerator(), true);
+        expect(pipelining.run()).to.be.be.rejectedWith('Error1');
+
+    });
+
+    it('should throw when null load generator is passed to setLoadGenerator', function () {
+        var pipelining = new Pipelining(1, fakeLoadGenerator());
+        expect(function () {
+            pipelining.setLoadGenerator(null);
+        }).to.throw(assert.AssertionError);
+    });
+
+    it('should run pipeline more than once without the storage of the results', function () {
+        var actual = [];
+
+        var pipelining = new Pipelining(100, createLoadGeneratorWithHandler(map, actual, 0));
+        return pipelining.run().then(function () {
+           pipelining.setLoadGenerator(createLoadGeneratorWithHandler(map, actual, actual.length));
+           return pipelining.run();
+        }).then(function (results) {
+            expect(results).to.be.an('undefined');
+            expect(actual.slice(0, ITEM_COUNT)).to.deep.equal(expected);
+            expect(actual.slice(ITEM_COUNT)).to.deep.equal(expected);
+        });
+    });
+
+    it('should run pipeline more than once with the storage of the results', function () {
+        var pipelining = new Pipelining(100, createLoadGenerator(map), true);
+        return pipelining.run().then(function (results) {
+            expect(results).to.deep.equal(expected);
+            pipelining.setLoadGenerator(createLoadGenerator(map));
+            return pipelining.run();
+        }).then(function (results) {
+            expect(results.slice(0, ITEM_COUNT)).to.deep.equal(expected);
+            expect(results.slice(ITEM_COUNT)).to.deep.equal(expected);
+        });
+    });
+
+    it('should not do more operations when the load generator is exhausted without the storage of results ', function () {
+        var actual = [];
+
+        var pipelining = new Pipelining(100, createLoadGeneratorWithHandler(map, actual, 0));
+        return pipelining.run().then(function (result) {
+            expect(result).to.be.an('undefined');
+            expect(actual).to.deep.equal(expected);
+            return pipelining.run();
+        }).then(function (result) {
+            expect(result).to.be.an('undefined');
+            expect(actual).to.deep.equal(expected);
+        })
+    });
+
+
+    it('should not do more operations when the load generator is exhausted with the storage of results ', function () {
+        var pipelining = new Pipelining(100, createLoadGenerator(map), true);
+        return pipelining.run().then(function (result) {
+            expect(result).to.deep.equal(expected);
+            return pipelining.run();
+        }).then(function (result) {
+            expect(result).to.deep.equal(expected);
+        })
+    });
+
```

Comment:
Can you write a test that actually verifies that "depth" is respected and there is not more than "depth" operations any given time? I know it is not trivial to write something like that.

Maybe we can write a test with a load generator that verifies internal in-flight requests counter of the pipelining object.

Reply:
A follow-up to this discussion.

I think that a running IMDG member isn't required for most of the tests in this module. As `Pipelining` isn't coupled with other objects in the client library, we can use any generator-like function that would return promises with `Promise.resolve`/`.reject`/`.delay`/etc. With this approach it'll be more or less trivial to implement a test that verifies depth limit guarantees.

So, I suggest to implement a separate suite of "quick" tests for pipelining. WDYT?

Reply:
+1 to testing Pipelining throughly in isolation. I still would like to see one or two tests involving some map operations to make sure they integrate well.